### DISCUSSION TITLE
RFC: use* functions may only be called during the `setup()` or other lifecycle hooks.

### DISCUSSION THREAD

**Discussion by franck-lynn** [urql-graphql/urql#2422] [Q&A] [answered]
## Summary

I want use useQuery in handClick method, but got the error.

use* functions may only be called during the `setup()` or other lifecycle hooks.

how can i solve it? thank. 
```
<script setup lang="ts">
import {provideClient, createClient} from '@urql/vue'
const client = createClient({
    url: "http://localhost:3000/graphql",
})
provideClient(client)
</script>
<template>
    <n-message-provider>
        <h1> Conde gen</h1>
        <query-urql-demo2></query-urql-demo2>
    </n-message-provider>
</template>
```

```
<script setup lang="ts">
import { useQuery, gql } from "@urql/vue"
import { onMounted, ref, watchEffect, watch, computed } from "vue"

const queryUsers = gql`
    query {
        queryUsers {
            _id
            name
        }
    }
` 
const data = ref()

const handleClick = () => {
    const res = useQuery({
        query: queryUsers,
    })
    data.value = res.data
    console.log(useQuery)
}
</script>
<template>
    <h3>query</h3>
    <n-button size="large" @click="handleClick"> queryUsers </n-button>
    <pre>
         {{ data }}
    </pre>
</template>
```

**Comment by genu** [comment]
I'm also seeing this using `nuxt3`

The error happens if I try to use a query in a middleware

**Comment by IZUNA894** [comment]
any solution to this ?

**Comment by kitten** [comment, marked as answer]
Hiya, just to clarify, since this has come up in other discussions. The APIs in `@urql/vue` may be used as part of Vue's Composition API: https://vuejs.org/guide/extras/composition-api-faq.html#what-is-composition-api

This has new consequences that are novel in Vue and have some limitations on it. When you use `useQuery` you basically are setting up a subscription to events and are receiving the results of it. This basically means that the `useQuery` calls is tied to your component's lifecycle. Until the component is unmounted, and as the inputs to the refs to this call change, the `Client` in `urql` will be able to update your query as the cache changes.

This means that you don't get a single result, but potentially many updating ones as demonstrated in the docs.

When you want to run a query programmatically you have instead two other options:
- If you want to update the UI and run a query when an interaction is performed, then "pausing" the `useQuery` subscription is a better choice: https://formidable.com/open-source/urql/docs/basics/vue/#pausing-usequery
- If you need only _a single query_ result, which mostly is more common with mutations, then you may also instead choose to use the promise-based API in `@urql/core` by getting the client with [`useClientHandle`](https://formidable.com/open-source/urql/docs/api/vue/#useclienthandle) and then running a `client.query(...).toPromise()` call: https://formidable.com/open-source/urql/docs/basics/core/#one-off-queries-and-mutations
